File: CF/Round597/d.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
    int n;
    scanf("%d", &n);
    vector<pair<int, int>> cities(n + 1);
    vector<int> k(n + 1);
    vector<long long> c(n + 1);
    set<int> visited;
    set<int> all;
    for (int i = 1; i <= n; i++) {
        all.insert(i);
    }
    set<int> built;
    vector<pair<int, int>> connections;
    for (int i = 1; i <= n; i++) {
        int x, y;
        scanf("%d%d", &x, &y);
        cities[i] = {x, y};
    }
    long long output = 0;
    int min_index = 0;
    long long min_value = LLONG_MAX;
    for (int i = 1; i <= n; i++) {
        scanf("%d", &c[i]);
        if (c[i] < min_value) {
            min_value = c[i];
            min_index = i;
        }
    }
    for (int i = 1; i <= n; i++) {
        scanf("%d", &k[i]);
    }
    visited.insert(min_index);
    built.insert(min_index);
    all.erase(min_index);
    output += min_value;
    for (int unvisited_city : all) {
        min_index = 0;
        min_value = LLONG_MAX;
        bool light;
        for (int visited_city : visited) {
            long long wire_cost = (long long)(k[visited_city] + k[unvisited_city]) * (long long)(abs(cities[visited_city].first - cities[unvisited_city].first) + abs(cities[visited_city].second - cities[unvisited_city].second));
            if (c[unvisited_city] < min_value) {
                min_value = c[unvisited_city];
                min_index = unvisited_city;
                light = true;
            }
            if (wire_cost < min_value) {
                min_value = wire_cost;
                min_index = visited_city;
                light = false;
            }
        }
        visited.insert(min_index);
        output += min_value;
        if (light) {
            built.insert(unvisited_city);
        } else {
            connections.push_back({unvisited_city, min_index});
        }
    }
    printf("%d\n", output);
    printf("%d\n", built.size());
    for (int city : built) {
        printf("%d ", city);
    }
    printf("\n");
    printf("%d\n", connections.size());
    for (pair<int, int> connection : connections) {
        printf("%d %d\n", connection.first, connection.second);
    }
}

File: CF/Round597/d.py
n = int(input())
unvisited = {i for i in range(n)}
visited = set()
lights = []
connections = []
cities = []
output = 0
for _ in range(n):
    x, y = [int(i) for i in input().split(' ')]
    cities.append((x, y))
c = [int(i) for i in input().split(' ')]
k = [int(i) for i in input().split(' ')]

min_index = 0
min_value = float('inf')
for i in range(n):
    if c[i] < min_value:
        min_value = c[i]
        min_index = i

visited.add(min_index)
unvisited.remove(min_index)
lights.append(min_index)
output += min_value

for u in unvisited:
    min_index = 0
    min_value = float('inf')
    light = False
    for v in visited:
        wire_cost = (k[v] + k[u]) * (abs(cities[v][0] - cities[u][0]) + abs(cities[v][1] - cities[u][1]))
        if c[u] < min_value:
            min_value = c[u]
            min_index = u
            light = True
        if wire_cost < min_value:
            min_value = wire_cost
            min_index = v
            light = False
    visited.add(u)
    output += min_value
    if light:
        lights.append(u)
    else:
        connections.append((min_index, u))
lights = [i + 1 for i in lights]
o = str(output) + "\n"
o += str(len(lights)) + "\n"
for i in lights:
    o += str(i) + " "
o += "\n"
o += str(len(connections)) + "\n"
for connection in connections:
    o += f'{connection[0] + 1} {connection[1] + 1}\n'
print(o)
